source/s3_auto_ranged_put.c

```diff
@@ -346,20 +351,38 @@ static void s_s3_meta_request_auto_ranged_put_destroy(struct aws_s3_meta_request
 
     aws_string_destroy(auto_ranged_put->synced_data.list_parts_continuation_token);
 
-    for (size_t checksum_index = 0; checksum_index < aws_array_list_length(&auto_ranged_put->encoded_checksum_list);
+    for (size_t checksum_index = 0;
+         checksum_index < aws_array_list_length(&auto_ranged_put->synced_data.encoded_checksum_list);
          ++checksum_index) {
 
         struct aws_byte_buf checksum_buf;
-        aws_array_list_get_at(&auto_ranged_put->encoded_checksum_list, &checksum_buf, checksum_index);
+        aws_array_list_get_at(&auto_ranged_put->synced_data.encoded_checksum_list, &checksum_buf, checksum_index);
 
         aws_byte_buf_clean_up(&checksum_buf);
     }
     aws_array_list_clean_up(&auto_ranged_put->synced_data.etag_list);
-    aws_array_list_clean_up(&auto_ranged_put->encoded_checksum_list);
+    aws_array_list_clean_up(&auto_ranged_put->synced_data.encoded_checksum_list);
     aws_http_headers_release(auto_ranged_put->synced_data.needed_response_headers);
     aws_mem_release(meta_request->allocator, auto_ranged_put);
 }
 
+/* Check flags and corresponding conditions to see if any more parts can be
+ * scheduled during this pass. */
+static bool s_should_skip_scheduling_more_parts_based_on_flags(
+    const struct aws_s3_auto_ranged_put *auto_ranged_put,
+    uint32_t flags) {
+    if ((flags & AWS_S3_META_REQUEST_UPDATE_FLAG_CONSERVATIVE) != 0) {
+        uint32_t num_parts_in_flight =
+            (auto_ranged_put->synced_data.num_parts_sent - auto_ranged_put->synced_data.num_parts_completed);
+
+        /* Because uploads must read from their streams serially, we try to limit the amount of in flight
+         * requests for a given multipart upload if we can. */
+        return num_parts_in_flight > 0;
```

Comment:
Isn't this going to limit us to only working on 1 part at a time? Like, we won't even start reading the 2nd part until the 1st part is 100% complete, having received an HTTP response?

That's going to be real slow. Once a part is done reading, we want to immediately start reading the next part, it's fine to have dozens of parts incomplete because they're still doing their HTTP transfer

Reply:
this is existing flow, i just pulled it into a separate function. 

our scheduler does 2 passes - first conservative and then regular. conservative pass always picks at most 1 request from each meta request to schedule. regular pass picks up as many requests as possible from meta request and then proceeds to next one